Add Jackson 3 support and deprecate Jackson 2 one

sdeleuze opened the pull request:
This PR adds support for Jackson 3 which has the following major differences with the Jackson 2 one:
 - `jackson` subpackage instead of `jackson2`
 - `Jackson` type prefix instead of `Jackson2`
 - `JsonMapper` instead of `ObjectMapper`
 - For configuration, `JsonMapper.Builder` instead of `ObjectMapper` since the latter is now immutable
 - `AllowlistTypeResolverBuilder` in new a public type in order to be used easily with the `JsonMapper.Builder` API

Jackson 3 changes compared to Jackson 2 are documented on https://github.com/FasterXML/jackson-future-ideas/discussions/72.

It also deprecates Jackson 2 support for removal.

rwinch reviewed (changes requested):
Thanks for the pull request. Generally, I think this is looking good. I've provided some feedback inline.

It also appears that there are still some classes that need migrated. Searching for jackson on main should give you a complete list of classes that are using jackson.

For example, [webauthn has jackson support](https://github.com/spring-projects/spring-security/tree/7.0.0-M2/webauthn/src/main/java/org/springframework/security/web/webauthn/jackson). I'm guessing this was missed because it is inconsistent with the rest of the jackson support in that it is in a jackson package rather than jackson2. It is also automatically registered because it does not use default types which would make it insecure.

There are also some tests that still use `ObjectMapper` or deprecated Spring Framework Jackson 2 classes (e.g. [JwtDecodersTests.java#L388](https://github.com/spring-projects/spring-security/blob/7.0.0-M2/oauth2/oauth2-jose/src/test/java/org/springframework/security/oauth2/jwt/JwtDecodersTests.java#L388)). Unless the test is specifically for the Jackson 2 support, we should update these to use non-deprecated classes (e.g. `JsonMapper`).


There are various Spring framework jackson based classes that have been deprecated that we should migrate away from. For example, `MappingJackson2HttpMessageConverter` usage (e.g. [WebAuthnAuthenticationFilter.converter](https://github.com/spring-projects/spring-security/blob/7.0.0-M2/webauthn/src/main/java/org/springframework/security/web/webauthn/authentication/WebAuthnAuthenticationFilter.java#L75)) should be migrated to `JacksonJsonHttpMessageConverter` if jackson 3 is on the classpath. Likely there could be a static factory method used that Spring Security uses to obtain the correct default json converter instance.

rwinch commented on core/src/main/java/org/springframework/security/jackson/AllowlistTypeResolverBuilder.java:
```diff
@@ -0,0 +1,161 @@
+ * @author Rob Winch
+ * @since 7.0
+ */
+public class AllowlistTypeResolverBuilder extends DefaultTypeResolverBuilder {
```
> From your comment, looks like Spring Security Jackson 3 modules would stop configuring AllowlistTypeResolverBuilder. But would we keep a Jackson 3 AllowlistTypeResolverBuilder or not (I think not but please confirm)?

No. We should remove Jackson 3 `AllowlistTypeResolverBuilder` so that we don't need to maintain that logic once we remove Jackson 2 support.

> Do you expect Jackson 3 mixins provided by Spring Security to use per class annotations instead?

Yes.

> I don't understand this sentence, could you please elaborate?

Sorry that this wasn't clear.

Previously, Spring Security's Jackson support forced enabling global default typing which is dangerous due to deserialization gadgets that might be exposed. Spring Security's Jackson 3 support should only enable per class (with classes that are safe). If users want global default typing, they will need to do that explicitly outside of Spring Security's Jackson 3 support.

> A raw draft of what the doc should say would be useful for my own understanding and for this PR, I will likely let you refine post-PR.

Expanding this statement, but keeping to a summary the docs would:
1) Explain that Spring Security's Jackson 3 support does not enable [global default typing](https://github.com/FasterXML/jackson-docs/wiki/JacksonPolymorphicDeserialization#11-global-default-typing) because it is [insecure](https://github.com/FasterXML/jackson-docs/wiki/JacksonPolymorphicDeserialization#111-security-risks-using-global-default-typing).
2) If users want to add additional classes, they can use [per-class annotations](https://github.com/FasterXML/jackson-docs/wiki/JacksonPolymorphicDeserialization#12-per-class-annotations). Alternatively, if users understand the security risks they can enable global default typing themselves.

rwinch commented on core/src/main/java/org/springframework/security/jackson/AllowlistTypeResolverBuilder.java:
```diff
@@ -0,0 +1,161 @@
+ * @author Rob Winch
+ * @since 7.0
+ */
+public class AllowlistTypeResolverBuilder extends DefaultTypeResolverBuilder {
```
Yes. I think that this should ultimately be removed.

rwinch commented on core/src/main/java/org/springframework/security/jackson/AbstractUnmodifiableCollectionDeserializer.java:
```diff
@@ -0,0 +1,68 @@
+ * @author Hyunmin Choi
+ * @since 7.0
+ */
+abstract class AbstractUnmodifiableCollectionDeserializer<T> extends ValueDeserializer<T> {
```
I think that this is it here:

* https://github.com/FasterXML/jackson-databind/issues/1881
* https://github.com/FasterXML/jackson-databind/issues/1880

sdeleuze commented:
@rwinch I have removed the custom support for unmodifiable collections (seems to work fine with Jackson 3), reworked the commit for a cleaner git history, and rebased the commits on `main`.

sdeleuze commented:
@jvalkeal Good catch, I will add a related commit.

sdeleuze commented:
@rwinch Rebasing is blocked on `FactorGrantedAuthority` not providing any public constructor but only a builder. It is possible to configure Jackson to use it, but supporting both Jackson 2 and 3 will require a bit more work, so before doing that I would like to double check with you that's expected.

sdeleuze commented:
After a discussion with @rwinch, we agreed on using [safe default typing](https://cowtowncoder.medium.com/jackson-2-10-safe-default-typing-2d018f0ce2ba) instead of disabling it totally. I will refine the PR accordingly.
